### PR TITLE
fix issue 20758 - __traits(getAttributes) as template argument result…

### PR DESCRIPTION
**Pull request by ghost**
…s in broken template instances

Using a `TupleExp` obtained by a `__traits` had for effect to instantiate a template with a `VarExp` that was never translated to a D symbol.
The fix is to do the translation only when the var is for a function, which means that the intention is not to instantiate using the function *result* but well the function itself.

### CODE REVIEW

File: src/dmd/dtemplate.d [hunk 1]
```diff
@@ -6515,7 +6515,16 @@ extern (C++) class TemplateInstance : ScopeDsymbol
 
                     if (ea.op == TOK.variable)
                     {
-                        /* This test is to skip substituting a const var with
+                        /* If the parameter is a function that is not called
+                         * explicitly then takes it a dsymbol.
```
Comment (thewilsonator) [inline src/dmd/dtemplate.d:6519]:
```
then it takes a dsymbol
```

**Review by thewilsonator — APPROVED** [review]
otherwise looks good

File: src/dmd/dtemplate.d (hunk 1 above)
Comment (NilsLankila) [reply src/dmd/dtemplate.d:6519]: I meant "as" I think

Comment (thewilsonator) [reply src/dmd/dtemplate.d]: This is still grammatically incorrect.

File: test/compilable/b20758.d [hunk 2]
```diff
@@ -0,0 +1,15 @@
+module b20758;
+
+template foo(A...) { }
+
+int attr() {return 1;}
+@attr int y;
+
+alias A = __traits(getAttributes, y);
+alias B = __traits(getOverloads, m, "attr");
```
Comment (NilsLankila) [inline test/compilable/b20758.d:9]: I forgot that the module was renamed after saving in dmd folder.

File: src/dmd/dtemplate.d (hunk 1 above)
Comment (NilsLankila) [reply src/dmd/dtemplate.d]: you mean "takes" ?

Comment (thewilsonator) [reply src/dmd/dtemplate.d]: "then it takes", not "then takes it"

Comment (NilsLankila) [reply src/dmd/dtemplate.d]: before I force push again, is this alright ?

```
If the parameter is a function that is not called explicitly then take it as a dsymbol.
```

Comment (thewilsonator) [reply src/dmd/dtemplate.d]: Hmm, looking at the code in context a bit more, probably:

```
If the parameter is a function that is not called explicitly, i.e. `foo!func` as opposed to `foo!func()`, then it is a dsymbol, not the return value of `func()`
```
is best.

**Pull request merged**